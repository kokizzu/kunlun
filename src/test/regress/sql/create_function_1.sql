--
-- CREATE_FUNCTION_1
--/home/kunlun/kunlun/src/test/regress

CREATE FUNCTION widget_in(cstring)
   RETURNS widget
   AS '//home/kunlun/kunlun/src/test/regress/regress.so'
   LANGUAGE C STRICT IMMUTABLE;

CREATE FUNCTION widget_out(widget)
   RETURNS cstring
   AS '//home/kunlun/kunlun/src/test/regress/regress.so'
   LANGUAGE C STRICT IMMUTABLE;

CREATE FUNCTION int44in(cstring)
   RETURNS city_budget
   AS '//home/kunlun/kunlun/src/test/regress/regress.so'
   LANGUAGE C STRICT IMMUTABLE;

CREATE FUNCTION int44out(city_budget)
   RETURNS cstring
   AS '//home/kunlun/kunlun/src/test/regress/regress.so'
   LANGUAGE C STRICT IMMUTABLE;

CREATE FUNCTION check_primary_key ()
	RETURNS trigger
	AS '//home/kunlun/kunlun/src/test/regress/refint.so'
	LANGUAGE C;

CREATE FUNCTION check_foreign_key ()
	RETURNS trigger
	AS '//home/kunlun/kunlun/src/test/regress/refint.so'
	LANGUAGE C;

CREATE FUNCTION autoinc ()
	RETURNS trigger
	AS '//home/kunlun/kunlun/src/test/regress/autoinc.so'
	LANGUAGE C;

CREATE FUNCTION trigger_return_old ()
        RETURNS trigger
        AS '//home/kunlun/kunlun/src/test/regress/regress.so'
        LANGUAGE C;

CREATE FUNCTION ttdummy ()
        RETURNS trigger
        AS '//home/kunlun/kunlun/src/test/regress/regress.so'
        LANGUAGE C;

CREATE FUNCTION set_ttdummy (int4)
        RETURNS int4
        AS '/home/kunlun/kunlun/src/test/regress/regress.so'
        LANGUAGE C STRICT;

CREATE FUNCTION make_tuple_indirect (record)
        RETURNS record
        AS '//home/kunlun/kunlun/src/test/regress/regress.so'
        LANGUAGE C STRICT;

CREATE FUNCTION test_atomic_ops()
    RETURNS bool
    AS '//home/kunlun/kunlun/src/test/regress/regress.so'
    LANGUAGE C;

-- Tests creating a FDW handler
CREATE FUNCTION test_fdw_handler()
    RETURNS fdw_handler
    AS '//home/kunlun/kunlun/src/test/regress/regress.so', 'test_fdw_handler'
    LANGUAGE C;

-- Things that shouldn't work:

CREATE FUNCTION test1 (int) RETURNS int LANGUAGE SQL
    AS 'SELECT ''not an integer'';';

CREATE FUNCTION test1 (int) RETURNS int LANGUAGE SQL
    AS 'not even SQL';

CREATE FUNCTION test1 (int) RETURNS int LANGUAGE SQL
    AS 'SELECT 1, 2, 3;';

CREATE FUNCTION test1 (int) RETURNS int LANGUAGE SQL
    AS 'SELECT $2;';

CREATE FUNCTION test1 (int) RETURNS int LANGUAGE SQL
    AS 'a', 'b';

CREATE FUNCTION test1 (int) RETURNS int LANGUAGE C
    AS 'nosuchfile';

CREATE FUNCTION test1 (int) RETURNS int LANGUAGE C
    AS '//home/kunlun/kunlun/src/test/regress/regress.so', 'nosuchsymbol';

CREATE FUNCTION test1 (int) RETURNS int LANGUAGE internal
    AS 'nosuch';
