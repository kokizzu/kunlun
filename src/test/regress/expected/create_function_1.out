--
-- CREATE_FUNCTION_1
--
CREATE FUNCTION widget_in(cstring)
   RETURNS widget
   AS '/home/jenkins/pgregressdata/lib/regress.so'
   LANGUAGE C STRICT IMMUTABLE;
psql:sql/create_function_1.sql:8: NOTICE:  type "widget" is not yet defined
DETAIL:  Creating a shell type definition.
CREATE FUNCTION
CREATE FUNCTION widget_out(widget)
   RETURNS cstring
   AS '/home/jenkins/pgregressdata/lib/regress.so'
   LANGUAGE C STRICT IMMUTABLE;
psql:sql/create_function_1.sql:13: NOTICE:  argument type widget is only a shell
CREATE FUNCTION
CREATE FUNCTION int44in(cstring)
   RETURNS city_budget
   AS '/home/jenkins/pgregressdata/lib/regress.so'
   LANGUAGE C STRICT IMMUTABLE;
psql:sql/create_function_1.sql:18: NOTICE:  type "city_budget" is not yet defined
DETAIL:  Creating a shell type definition.
CREATE FUNCTION
CREATE FUNCTION int44out(city_budget)
   RETURNS cstring
   AS '/home/jenkins/pgregressdata/lib/regress.so'
   LANGUAGE C STRICT IMMUTABLE;
psql:sql/create_function_1.sql:23: NOTICE:  argument type city_budget is only a shell
CREATE FUNCTION
CREATE FUNCTION check_primary_key ()
	RETURNS trigger
	AS '/home/jenkins/pgregressdata/lib/refint.so'
	LANGUAGE C;
CREATE FUNCTION
CREATE FUNCTION check_foreign_key ()
	RETURNS trigger
	AS '/home/jenkins/pgregressdata/lib/refint.so'
	LANGUAGE C;
CREATE FUNCTION
CREATE FUNCTION autoinc ()
	RETURNS trigger
	AS '/home/jenkins/pgregressdata/lib/autoinc.so'
	LANGUAGE C;
CREATE FUNCTION
CREATE FUNCTION trigger_return_old ()
        RETURNS trigger
        AS '/home/jenkins/pgregressdata/lib/regress.so'
        LANGUAGE C;
CREATE FUNCTION
CREATE FUNCTION ttdummy ()
        RETURNS trigger
        AS '/home/jenkins/pgregressdata/lib/regress.so'
        LANGUAGE C;
CREATE FUNCTION
CREATE FUNCTION set_ttdummy (int4)
        RETURNS int4
        AS '/home/jenkins/pgregressdata/lib/regress.so'
        LANGUAGE C STRICT;
CREATE FUNCTION
CREATE FUNCTION make_tuple_indirect (record)
        RETURNS record
        AS '/home/jenkins/pgregressdata/lib/regress.so'
        LANGUAGE C STRICT;
CREATE FUNCTION
CREATE FUNCTION test_atomic_ops()
    RETURNS bool
    AS '/home/jenkins/pgregressdata/lib/regress.so'
    LANGUAGE C;
CREATE FUNCTION
-- Tests creating a FDW handler
CREATE FUNCTION test_fdw_handler()
    RETURNS fdw_handler
    AS '/home/jenkins/pgregressdata/lib/regress.so', 'test_fdw_handler'
    LANGUAGE C;
CREATE FUNCTION
-- Things that shouldn't work:
CREATE FUNCTION test1 (int) RETURNS int LANGUAGE SQL
    AS 'SELECT ''not an integer'';';
psql:sql/create_function_1.sql:74: ERROR:  return type mismatch in function declared to return integer
DETAIL:  Actual return type is text.
CONTEXT:  SQL function "test1"
CREATE FUNCTION test1 (int) RETURNS int LANGUAGE SQL
    AS 'not even SQL';
psql:sql/create_function_1.sql:77: ERROR:  syntax error at or near "not"
LINE 2:     AS 'not even SQL';
                ^
CREATE FUNCTION test1 (int) RETURNS int LANGUAGE SQL
    AS 'SELECT 1, 2, 3;';
psql:sql/create_function_1.sql:80: ERROR:  return type mismatch in function declared to return integer
DETAIL:  Final statement must return exactly one column.
CONTEXT:  SQL function "test1"
CREATE FUNCTION test1 (int) RETURNS int LANGUAGE SQL
    AS 'SELECT $2;';
psql:sql/create_function_1.sql:83: ERROR:  there is no parameter $2
LINE 2:     AS 'SELECT $2;';
                       ^
CREATE FUNCTION test1 (int) RETURNS int LANGUAGE SQL
    AS 'a', 'b';
psql:sql/create_function_1.sql:86: ERROR:  only one AS item needed for language "sql"
CREATE FUNCTION test1 (int) RETURNS int LANGUAGE C
    AS 'nosuchfile';
psql:sql/create_function_1.sql:89: ERROR:  could not access file "nosuchfile": No such file or directory
CREATE FUNCTION test1 (int) RETURNS int LANGUAGE C
    AS '/home/jenkins/pgregressdata/lib/regress.so', 'nosuchsymbol';
psql:sql/create_function_1.sql:92: ERROR:  could not find function "nosuchsymbol" in file "/home/jenkins/pgregressdata/lib/regress.so"
CREATE FUNCTION test1 (int) RETURNS int LANGUAGE internal
    AS 'nosuch';
psql:sql/create_function_1.sql:95: ERROR:  there is no built-in function named "nosuch"
